qubesadmin/tools/qvm_shutdown.py

```diff
@@ -48,37 +48,66 @@
     help='timeout after which domains are killed when using --wait'
         ' (default: %d)')
 
+def distance_from_netvm(vm):
+    ''' Calculate distance from netvm chain root.
+
+    VM not connected to any netvm get 0, a VM connected to it get 1, etc.
+    '''
+    distance = 0
+    try:
+        while vm.netvm is not None:
+            distance += 1
+            vm = vm.netvm
+    except (KeyError, AttributeError):
+        # no netvm, or we don't have access to it
+        pass
+    return distance
 
 def main(args=None, app=None):  # pylint: disable=missing-docstring
     args = parser.parse_args(args, app=app)
 
-    for vm in args.domains:
-        try:
-            vm.shutdown(force=args.force)
-        except qubesadmin.exc.QubesVMNotStartedError:
-            pass
-
-    if not args.wait:
-        return
-
-    timeout = args.timeout
-    current_vms = list(sorted(args.domains))
-    while timeout >= 0:
-        current_vms = [vm for vm in current_vms
-            if vm.get_power_state() != 'Halted']
-        if not current_vms:
-            return 0
-        args.app.log.info('Waiting for shutdown ({}): {}'.format(
-            timeout, ', '.join([str(vm) for vm in current_vms])))
-        time.sleep(1)
-        timeout -= 1
-
-    args.app.log.info(
-        'Killing remaining qubes: {}'
-        .format(', '.join([str(vm) for vm in current_vms])))
-    for vm in current_vms:
-        vm.kill()
-
+    # care about order only when --wait is used:
+    if args.wait:
+        domains_and_distance = [(vm, distance_from_netvm(vm))
```

Comment:
I don't think this is right, since extensions can introduce other dependencies (like device attached for example) and disallow vm shutdown for that (or other) reason. This code assumes that the only dependency between vms is expressed as netvms.

I'd write it as (this is only a rough idea):
```python
@asyncio.coroutine
def shutdown_vm(vm):
  yield from vm.shutdown()
  return vm

for i in range(max_rounds): # maybe 7
  if not vms_to_shutdown:
    break
  done, pending = yield from asyncio.wait((shutdown_vm(vm) for vm in vms_to_shutdown), timeout=20)
  # (or whatever timeout is deemed OK)
  # TODO handle pending and don't call vm.shutdown() more than once
  for fut in done:
    try:
      vms_to_shutdown.remove(fut.result())
    except QubesException:  # if vm.shutdown() raises exception, it gets re-raised by fut.result()
      pass

# after max_rounds, they are fair game
if vms_to_shutdown:
  yield from asyncio.wait(vm.kill() for vm in vms_to_shutdown)
```

That way we allow for more flexible ordering.